### DISCUSSION TITLE
Support display on Lilygo T-QT (ESP32S3) - GC9107

### DISCUSSION THREAD

**Discussion by eni23** [Bodmer/TFT_eSPI#2175] [New processor/display support or capability request]
In order to get the display on a [Lilygo T-QT](https://www.lilygo.cc/products/t-qt-v1-1) working with TFT_eSPI, there is a change in `TFT_Drivers/GC9A01_Rotation.h` needed, otherwise there is a weird border and the display does not really displays the whole area.

The changed file can be found [here](https://github.com/Xinyuan-LilyGO/T-QT/blob/main/lib/TFT_eSPI/TFT_Drivers/GC9A01_Rotation.h).

It would be nice to integrate this change in TFT_eSPI upstream somehow so there won't be any need for maintaining a separate repo. Since just changing the file would probably break something else, i'm opening this issue here to ask how it could be done.

Thanks :)

**Comment by Bodmer** [comment]
The module seems to have a new display controller type GC9107. This appears to have the same commands and register init values as the GC9A01 but has different offsets since not all the pixel capability is used.  I think this ideally needs another driver file set to avoid confusion. e.g. #define GC9107_DRIVER. This is simple to implement and I may be able find time to update the library the near future. In the meantime I see you have a solution.

I assume the display works correctly in all possible rotations 0-3 with the proposed setup file? Correct?

**Comment by eni23** [comment]
> I assume the display works correctly in all possible rotations 0-3 with the proposed setup file? Correct?

Yes, all rotations work fine.

> I think this ideally needs another driver file set to avoid confusion. e.g. #define GC9107_DRIVER. This is simple to implement and I may be able find time to update the library the near future

If it is just adding the files `TFT_Drivers/GC9107_Init.h`,  `TFT_Drivers/GC9107_Rotation.h`,  `TFT_Drivers/GC9107_Defines.h` and a user-setup (which number should i choose?) and adjusting them accordingly, i can prepare a MR as well.

One thing i was wondering: Lilygo has a different display example whrere they do not need your library but a "custom driver". There the init commands look quite different than the ones in `GC9A01`: https://github.com/Xinyuan-LilyGO/T-QT/blob/main/example/lcd/lcd_driver.cpp#L109

But the ones from `GC9A01` seem to work perfectly fine so i assume just keep going with those ones?

**Comment by Bodmer** [comment]
It is best to use the specific setup recommended b the vendor to get the best out of the display in terms of colour fidelity and Gamma balance. The Lilygo code is useful as it can be adapted.

The driver files for init and rotation need to be called up in TFT_eSPI.cpp as well, the location is place marked circa line 700 and line 800. Also in User_Setup_Select.h circa line 200.

**Comment by mrudick1** [comment]
Bodmer:  Thank you for all the work you've done and continue to do on the TFT libraries.  I use them in a T-QT Pro project I've developed.  I concur that it would be very helpful to include a T-QT Pro driver.

eni23:  Thank you for solving this problem.  I had started to go a little crazy after having updated from 2.4.61 to 2.5.  Thankfully my head of hair remains..  for now.

**Comment by tochinet** [comment]
Hi eni23, bodmer, mrudick1, all,
Did any of you finalize the support for the T-QT (family) ? I didn't see it in the lib.
Best regards,
Tochinet

**Comment by Bodmer** [comment]
The display driver is not supported but according to this post the ST7789 driver works: #2201 

Note that a screen mirror was reported so the ST7789 driver might need to be adapted.

If @eni23 or @mrudick1 can create and test a driver and create a pull request, then I can add it.

**Reply by mrudick1** [reply]
I took a look at the documentation for both the GC9107 and GC9A01.  There doesn't seem to be any mirroring issue, as they both use the same GRAM bit settings.
In the LilyGo branch, there is now a new [User_Setup.h ](https://github.com/Xinyuan-LilyGO/T-QT/blob/main/lib/TFT_eSPI/User_Setups/Setup211_LilyGo_T_QT_Pro_S3.h)file defined, but it uses the GC9A01 driver.  That said, they also updated the [GC9A01_Rotation.h](https://github.com/Xinyuan-LilyGO/T-QT/blob/main/lib/TFT_eSPI/TFT_Drivers/GC9A01_Rotation.h) file.  Swapping them both out works with the T-QT Pro.
That said, it's not an entirely new driver, which would be cleaner.

**Reply by tochinet** [reply]
Thanks for the fast reply. I didn't get the boards yet, but I'll let you know if there are any issue.

**Comment by taxze6** [comment]
Hi all, I'm sorry because I'm a beginner, so after reading all's answers I still don't know how to use tft-eSPI to drive GC9107, I also use ESP32S3. After reading all the answers, do I understand that just changing `TFT_Drivers/ gc9a01_rotate.h` can migrate GC9A01 to GC9107? Can you give me some more advice? Thanks :)  This is very important to me 🥹

Sorry all, I have checked the project carefully again, and I found a commit "Added LilyGo T-QT Support" in `GC9A01_Rotation.h`. Does this mean that tft_eSPI has supported GC9107? My screen won't be available for a week, so I'll start with the technical implementation~
